apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -1541,6 +1655,128 @@ impl Telemetry {
+        let typename = opt_type.as_str().unwrap_or("");
+
+        res.push((Arc::new(typename.to_string()), representation.clone()));
+    }
+    Ok(res)
+}
+
+struct CacheCounter {
+    potential_cache_entries: CountMinSketch<CacheKey, usize>,
+    // By subgraph
+    nb_requests: HashMap<Arc<String>, usize>,
+    // By subgraph and typename
+    potential_cache_hit: HashMap<(Arc<String>, Arc<String>), usize>,
+    created_at: Instant,
+    ttl: Duration,
+}
+
+impl CacheCounter {
+    fn new(ttl: Duration) -> Self {
+        Self {
+            potential_cache_entries: CountMinSketch::new(
+                0.01, // probability calculate k_num
+                0.01, // error tolerance to calculate width
+                (),
+            ),
+            potential_cache_hit: HashMap::new(),
+            nb_requests: HashMap::new(),
+            created_at: Instant::now(),
+            ttl,
+        }
+    }
+
+    fn record(
+        &mut self,
+        query: Arc<String>,
+        subgraph_name: Arc<String>,
+        hashed_headers: Arc<String>,
+        representations: Vec<(Arc<String>, Value)>,
+    ) {
+        if self.created_at.elapsed() >= self.ttl {
+            self.clear();
+        }
+
+        // I record the total number of requests (asking for _entities) to a specific subgraph
+        // in order to know the frequency of potential cache hit we could have for a subgraph
+        let nb_requests = *self
+            .nb_requests
+            .entry(subgraph_name.clone())
+            .and_modify(|e| *e += 1)
+            .or_insert(1);
+
+        for (typename, representation) in representations {
+            let cache_hit = self.potential_cache_entries.push(
+                &CacheKey {
+                    representation,
+                    typename: typename.clone(),
+                    query: query.clone(),
+                    subgraph_name: subgraph_name.clone(),
+                    hashed_headers: hashed_headers.clone(),
+                },
+                &1,
+            ) > 1;
+            // I record the number of potential cache hit we would have for a subgraph and an entity
+            let potential_cache_hit = *self
+                .potential_cache_hit
+                .entry((subgraph_name.clone(), typename.clone()))
+                .and_modify(|e| {
+                    if cache_hit {
+                        *e += 1;
+                    }
+                })
+                .or_insert(1);
+
+            // Here I'm computing the percentage of potential cache hit I could have for a subgraph given its entity name
+            ::tracing::info!(
+                value.apollo.router.operations.entity = (potential_cache_hit as f64 / nb_requests as f64) * 100f64,
+                entity_type = %typename,
+                subgraph = %subgraph_name,
+            );
+        }
+    }
+
+    fn clear(&mut self) {
+        self.potential_cache_entries = CountMinSketch::new(
```

Comment:
If we're restarting from a new structure every time, resetting everything, cache misses would be overrepresented. What we can do here is make a kind of sliding window by alternating between 2 CountMinSketch structures:
- when we see an entity come in, we insert in the primary structure, and we get the count from both the primary and secondary structure (but without updating the count for the secondary)
- when we reach the TTL, we reset the secondary structure, and switch them

That way, if an entity was seen in the previous primary, it would still be seen as a cache hit, and then recorded in the new primary